dask/blockwise.py

```diff
@@ -591,7 +585,10 @@ def _dict(self):
 
             # Handle IO Subgraph
             for k in dsk:
-                io_key = (self.io_name,) + tuple([k[i] for i in range(1, len(k))])
+                # Leave out `new_axes` in key
+                # TODO: Check that this makes sense
+                sz = len([i for i in self.output_indices if i not in self.new_axes])
+                io_key = (self.io_name,) + tuple([k[i + 1] for i in range(sz)])
```

Comment:
@ian-r-rose - It seem that this change helps with [the problem you found](https://github.com/dask/dask/pull/6931#issuecomment-738972314).  However, I am not yet sure if this will always produce the correct IO key to replace.

Reply:
Thanks for looking into it @rjzamora. This seems to help with the original new-axes issue:
```python
def f(x):
    return x[:, None] * np.ones((1, 7))

x = da.ones(5, chunks=2)
y = da.blockwise(
  f, "aq", x, "a", new_axes={"q": 7}, concatenate=True, dtype=x.dtype
)
y.compute()
```
However, it fails again with the same operation transposed (again derived from the test suite):
```python
def f(x):
    return x[None, :] * np.ones((7, 1))

x = da.ones(5, chunks=2)
y = da.blockwise(
  f, "qa", x, "a", new_axes={"q": 7}, concatenate=True, dtype=x.dtype
)
y.compute()
```
<details>

```pytb
<ipython-input-3-988e4cf91f2a> in f(x)
      1 def f(x):
----> 2     return x[None, :] * np.ones((7, 1))
      3 
      4 x = da.ones(5, chunks=2)
      5 y = da.blockwise(

TypeError: 'cytoolz.functoolz.curry' object is not subscriptable
```

</details>

I'm puzzled as to what is going on such that the blockwise `f(x)` is getting fed a `curry` object rather than an array object, and how it relates to your logic in 4a90128. Could you elaborate a bit?

For what it's worth, these issues seem to disappear when skipping the [blockwise fusion](https://github.com/dask/dask/blob/3a08bb32e736d51a758281a3abc0fc0dbec04efc/dask/array/optimization.py#L46) step while optimization.

Reply:
>I'm puzzled as to what is going on such that the blockwise f(x) is getting fed a curry object rather than an array object, and how it relates to your logic in 4a90128. Could you elaborate a bit?

My understanding (which could be off), is that this that the problems are related to the requirement that `BlockwiseIO` remove place-holder keys with the the actual IO functions they refer to when the low-level graph is materialized.  **More specifically**: What we are ultimately doing for your `BlockwiseCreateArray` case is "pretending" that we have an existing `dask.array` with the name `"blockwise-create-..."`, but replacing each key name with the actual IO function in the final `subgraph_callable` object.  This is working fine for a standalone `BlockwiseIO` layer, but seems to fail after we fuse together two `Blockwise` layers with misaligned indices.

>For what it's worth, these issues seem to disappear when skipping the blockwise fusion step while optimization.

Exactly right.  The problem occurs when we fuse misaligned indices, and then we fail to remove the place-holder "IO" keys from the graph.

Reply:
I pushed more changes to handle the case that the indices are reordered between adjacent layers.  So now, your second example should work (I hope).  Thanks for your patience here @ian-r-rose !